pkg/driver/microphone/microphone.go

```diff
@@ -136,11 +141,35 @@ func (m *microphone) AudioRecord(inputProp prop.Media) (audio.Reader, error) {
 		return nil, err
 	}
 
+	var closeDeviceOnce sync.Once
+	m.deviceCloseFunc = func() {
+		closeDeviceOnce.Do(func() {
+			closeDone := make(chan struct{})
+			// we may be waiting on a chunk so keep reading until we can stop
+			// the malgo device.
+			go func() {
+				for {
+					select {
+					case <-closeDone:
+						return
+					default:
+					}
+					select {
+					case <-m.chunkChan:
+					case <-closeDone:
+					}
```

Comment:
Is `<-m.chunkChan` always received after `<-closeDone`?
I guess `m.chunkChan` and `closeDone` may be triggered almost at the same time and may return before consuming `m.chunkChan` at low chance.

```go
func (m *microphone) Close() error {
	if m.deviceCloseFunc != nil {
		m.deviceCloseFunc()
	}
	// Don't close m.chunkChan to avoid panic due to `send to closed chan`
	// Go GC collects unclosed chan with buffered value when unused:
	//   https://groups.google.com/g/golang-nuts/c/KtIyc5lTRJY/m/6PK6aK5QKHIJ
	return nil
}
...
func (m *microphone) AudioRecord(inputProp prop.Media) (audio.Reader, error) {
	...
	ctx, cancel := context.WithCancel(context.Background())
	onRecvChunk := func(_, chunk []byte, framecount uint32) {
		select {
		case m.chunkChan <- chunk:
		case <-ctx.Done():
		}
	}
	...
	m.deviceCloseFunc = func() {
		closeDeviceOnce.Do(func() {
			cancel() // Unblock onRecvChunk
			device.Stop()
			device.Uninit()
		})
	}
```
would be more safe?

Reply:
Yup, good catch. Will use your version. The whole point is to unblock anyway, not consume.

Reply:
Also check your gophers slack!

Reply:
I'll still use your way but the original way is also safe because closeDone only happens after we stop the device which means the callback will never fire and no one will get stuck.